base_layer/core/src/base_node/states/forward_block_sync.rs

```diff
@@ -0,0 +1,278 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    base_node::{
+        states::{ListeningInfo, StateEvent},
+        BaseNodeStateMachine,
+    },
+    blocks::{BlockHash, BlockHeader},
+    chain_storage::{async_db, BlockchainBackend, BlockchainDatabase, ChainMetadata, ChainStorageError},
+    transactions::types::HashOutput,
+};
+use log::*;
+use rand::{rngs::OsRng, Rng};
+use std::{cmp, collections::VecDeque};
+use tari_comms::peer_manager::NodeId;
+use tari_crypto::tari_utilities::{hex::Hex, Hashable};
+
+const LOG_TARGET: &str = "c::bn::states::block_sync";
+
+// The maximum number of retry attempts a node can perform to request a particular block from remote nodes.
+const MAX_HEADER_REQUEST_RETRY_ATTEMPTS: usize = 5;
+const MAX_BLOCK_REQUEST_RETRY_ATTEMPTS: usize = 5;
+const MAX_HEADER_HASHES_TO_SEND: u64 = 128;
+const MAX_BLOCKS_TO_DOWNLOAD: usize = 20;
```

Comment:
This will work for empty blocks, but will be problematic when the blocks are full as it will result in a response message that is +-20MBytes.